anonlink/serialization.py

```diff
@@ -0,0 +1,221 @@
+
+# https://docs.python.org/3/library/array.html#module-array
+_ARRAY_UINT_TYPES = 'BIHLQ'
+_ARRAY_FLOAT_TYPES = 'df'
+_ARRAY_UINT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                          for t in _ARRAY_UINT_TYPES}
+_ARRAY_FLOAT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                           for t in _ARRAY_FLOAT_TYPES}
+
+_HEADER_STRUCT = _struct.Struct('<BBBB')
+
+
+_CandidatePair = _typing.Tuple[float, int, int, int, int]
+_CandidatePairIter = _typing.Iterable[_CandidatePair]
+
+# This is subject to change.
+# The format is composed of a header and a sequence of entries.
+# The header is composed of:  version (1 byte), number of bytes in
+# similarity (1 byte), number of bytes in dataset index (1 byte), number
+# of bytes in record index (1 byte).
+# Every entry is composed of a similarity as a IEEE floating point
+# value, a two datset indices, and two record indices. Their sizes as
+# determined by the header.
+# All integers are unsigned. All values are *little-endian* (yes, this
+# breaks convention). The number of candidate matchings is not included
+# on purpose: if needed, it can be computed from the length of the file.
+
+
+def _entry_struct(
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int
+) -> _struct.Struct:
+    try:
+        sim_t = _STRUCT_FLOAT_LEN_TO_FMT[sim_t_size]
+    except KeyError:
+        msg = f'floats of {sim_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        dset_i_t = _STRUCT_UINT_LEN_TO_FMT[dset_i_t_size]
+    except KeyError:
+        msg = f'indices of {dset_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    try:
+        rec_i_t = _STRUCT_UINT_LEN_TO_FMT[rec_i_t_size]
+    except KeyError:
+        msg = f'indices of {rec_i_t_size} bytes are not supported'
+        raise ValueError(msg) from None
+    return _struct.Struct(f"<{sim_t}2{dset_i_t}2{rec_i_t}")
+    
+
+def _dump_from_iterable(
+    f: _typing.BinaryIO,
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int,
+    iterable: _CandidatePairIter
+) -> None:
+    # Fail without writing if the parameters are incorrect.
+    entry_struct = _entry_struct(
+        sim_t_size, dset_i_t_size, rec_i_t_size)
+    
+    # Write header.
+    f.write(_HEADER_STRUCT.pack(1, sim_t_size, dset_i_t_size, rec_i_t_size))
+    
+    for entry in iterable:
+        f.write(entry_struct.pack(*entry))
+
+
+def _entries_iterable(
+    f: _typing.BinaryIO,
+    entry_struct: _struct.Struct
+) -> _CandidatePairIter:
+    buffer = f.read(entry_struct.size)
+    while buffer:
+        if len(buffer) != entry_struct.size:
+            raise ValueError('ran out of input')
+        yield _typing.cast(_CandidatePair, entry_struct.unpack(buffer))
+        buffer = f.read(entry_struct.size)
+
+
+def _load_to_iterable(
+    f: _typing.BinaryIO
+) -> _typing.Tuple[_CandidatePairIter, int, int, int]:
+    # Make sure that f.read(n) returns exactly n bytes.
```

Comment:
As far as I know [`BufferedIO`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read) doesn't guarantee this property of returning exactly `n` bytes? Although in practice - with small `n` - it does.

Reply:
From [the docs](https://docs.python.org/3/library/io.html#io.BufferedIOBase):
> The main difference with `RawIOBase` is that methods `read()`, `readinto()` and `write()` will try (respectively) to read as much input as requested or to consume all given output, at the expense of making perhaps more than one system call.

It looks like [there is an exception](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read):
> But for interactive raw streams, at most one raw read will be issued, and a short result does not imply that EOF is imminent.

The [implementation](https://github.com/python/cpython/blob/55edd0c185ad2d895b5d73e47d67049bc156b654/Modules/_io/bufferedio.c#L881) of `BufferedIO.read(n)` confirms this. It reads the input until `n` bytes are read (or until the raw stream is exhausted for `n == -1`).

Reply:
Interesting, looks like the Python docs are a bit ambiguous here. Good work checking up